File: WordSearch.py
from typing import List

'''
LeetCode 79. Word Search
'''

class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rowlen, collen = len(board), len(board[0])
        visited = set()

        def wordsearch_backtrack(row:int, col:int, charPos:int)->bool:
            if charPos >= len(word):
                return True
            print(row, col, charPos)
            
            for i,j in [[row-1, col], [row+1, col], [row, col-1], [row, col+1]]:
                if i >=0 and i < rowlen and j >= 0 and j < collen and board[i][j] == word[charPos] and (i,j) not in visited:
                    visited.add((i, j))
                    if wordsearch_backtrack(i, j, charPos+1):
                        return True
                    visited.remove((i, j))
            return False

        
        for i in range(rowlen):
            for j in range(collen):
                if board[i][j] == word[0]:
                    visited.add((i, j))
                    if wordsearch_backtrack(i, j, 1):
                        return True
                    else:
                        visited.remove((i, j))
        return False
    


sol = Solution()
#print(sol.exist([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "ABCCED"))
#print(sol.exist([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "SEE"))
#print(sol.exist([["A","B","C","E"],["S","F","C","S"],["A","D","E","E"]], "ABCB"))
print(sol.exist([["A","A"]], "AAA"))

File: WordSearch.java
import java.util.*;

public class WordSearch {

    int[][] visited;
    int rowlen;
    int collen;

    public boolean isValid(List<Integer> move)
    {
      if(move.get(0) >= 0 && move.get(1) >= 0 && move.get(0) < rowlen && move.get(1) < collen && visited[move.get(0)][move.get(1)] == 0)
        return true;
    
       return false;
    }
    public boolean dfsBacktrack(char[][] board, String word, int row, int col, int idx)
    {
        if(idx >= word.length()-1)
            return true;
        
        List<List<Integer>> nextMoves = new ArrayList<>();
        nextMoves.add(Arrays.asList(new Integer[]{row-1, col}));
        nextMoves.add(Arrays.asList(new Integer[]{row+1, col}));
        nextMoves.add(Arrays.asList(new Integer[]{row, col+1}));
        nextMoves.add(Arrays.asList(new Integer[]{row, col-1}));

        //System.out.println(row + "," + col + "," + idx);

        for(List<Integer> curMove: nextMoves)
        {
            int r = curMove.get(0);
            int c = curMove.get(1);
            if(isValid(curMove) && board[r][c] == word.charAt(idx+1)){

                visited[r][c] = 1;
                if(dfsBacktrack(board, word, r, c, idx+1))
                    return true;
                
                visited[r][c] = 0;
            }
        }

        return false;
    }

    public boolean exist(char[][] board, String word) {
        word = new StringBuilder(word).reverse().toString();

        rowlen = board.length;
        collen = board[0].length;
        visited = new int[rowlen][collen];

        for(int i=0;i<rowlen;i++)
        {
            for(int j=0;j<collen;j++)
            {
                if(board[i][j] == word.charAt(0)){
                    visited[i][j] = 1;
                    if(dfsBacktrack(board, word, i, j, 0))
                        return true;

                    visited[i][j] = 0;
                }
            }
        }
        return false;
    }

    public static void main(String[] args)
    {
        WordSearch obj = new WordSearch();
        //char[][] board = {{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}};
        //char[][] board = {{'A','A'}};
        char[][] board = {{'A','A','A','A','A','A'},{'A','A','A','A','A','A'},{'A','A','A','A','A','A'},{'A','A','A','A','A','A'},{'A','A','A','A','A','B'},{'A','A','A','A','B','A'}};
        String word = "AAAAAAAAAAAAABB";
        System.out.println(obj.exist(board, word));
    }
}
